File: modules/napplet-host/ios/CapabilityLeaseRegistry.swift
import Foundation

/// Synchronous native authority. All mutable state is protected by the same lock;
/// no UIKit/WebKit reference or user key is stored here.
final class CapabilityLeaseRegistry: @unchecked Sendable {
  static let maxBytes = 2 * 1024 * 1024
  private struct Session { var sequence: UInt64 = 0 }
  private struct Request {
    let generation: String
    let snapshot: String
    let expires: ContinuousClock.Instant
    var claimed = false
  }
  private let lock = NSLock()
  private let now: @Sendable () -> ContinuousClock.Instant
  private var sessions: [String: Session] = [:]
  private var requests: [String: Request] = [:]

  init(now: @escaping @Sendable () -> ContinuousClock.Instant = { ContinuousClock.now }) {
    self.now = now
  }
  func register(_ generation: String) -> Bool {
    lock.withLock {
      guard generation.range(of: "^[A-Za-z0-9_-]{1,80}$", options: .regularExpression) != nil,
        sessions[generation] == nil, sessions.count < 64 else { return false }
      sessions[generation] = Session()
      return true
    }
  }
  func sessionActive(_ generation: String) -> Bool { lock.withLock { sessions[generation] != nil } }
  func revoke(_ generation: String) {
    lock.withLock {
      sessions.removeValue(forKey: generation)
      requests = requests.filter { $0.value.generation != generation }
    }
  }
  /// Snapshot is complete JSON constructed by the host from native registration and the original serialized request.
  func admit(_ generation: String, sequence: UInt64, snapshot: String) -> String? {
    lock.withLock {
      prune()
      guard var session = sessions[generation], sequence > 0,
        sequence <= 9_007_199_254_740_991, sequence == session.sequence + 1 else { return nil }
      // Overload consumes its transport sequence; a dropped request cannot become a later replay.
      session.sequence = sequence
      sessions[generation] = session
      guard snapshot.utf8.count <= Self.maxBytes, requests.count < 8,
        requests.values.filter({ $0.generation == generation }).count < 4 else { return nil }
      let token = UUID().uuidString.lowercased()
      requests[token] = Request(generation: generation, snapshot: snapshot, expires: now().advanced(by: .seconds(25)))
      return token
    }
  }
  func take(_ token: String) -> String? {
    lock.withLock {
      prune()
      guard var request = requests[token], !request.claimed, sessions[request.generation] != nil else { return nil }
      request.claimed = true
      requests[token] = request
      return request.snapshot
    }
  }
  func isActive(_ token: String) -> Bool {
    lock.withLock {
      prune()
      guard let request = requests[token] else { return false }
      return request.claimed && sessions[request.generation] != nil
    }
  }
  /// Consumes exactly this request. Delivery must still recheck its generation on the main actor.
  func finish(_ token: String) -> Bool {
    lock.withLock {
      prune()
      guard let request = requests.removeValue(forKey: token) else { return false }
      return request.claimed && sessions[request.generation] != nil
    }
  }
  private func prune() {
    let current = now()
    requests = requests.filter { current < $0.value.expires && sessions[$0.value.generation] != nil }
  }
}

File: modules/napplet-host/android/src/main/java/org/nostrocket/hypergolic/host/CapabilityLeaseRegistry.java
package org.nostrocket.hypergolic.host;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.function.LongSupplier;

/** Synchronous native authority, independent of delayed React Native events.
 * The host validates origin/frame, schema and immutable registration before admission. */
public final class CapabilityLeaseRegistry {
  public static final int MAX_BYTES = 2 * 1024 * 1024;
  private static final int MAX_SESSIONS = 64;
  private static final int MAX_PENDING = 8;
  private static final int MAX_PER_SESSION = 4;
  private static final long MAX_SEQUENCE = 9_007_199_254_740_991L;
  private static final long LIFETIME_MS = 25_000L;
  private final LongSupplier elapsedRealtime;
  private final Map<String, Session> sessions = new HashMap<>();
  private final Map<String, Request> requests = new HashMap<>();

  private static final class Session { long sequence = 0; }
  private static final class Request {
    final String generation;
    final String snapshot;
    final long expires;
    boolean claimed = false;
    Request(String generation, String snapshot, long expires) {
      this.generation = generation; this.snapshot = snapshot; this.expires = expires;
    }
  }
  /** Android supplies SystemClock.elapsedRealtime, which includes device sleep. */
  public CapabilityLeaseRegistry(LongSupplier elapsedRealtime) { this.elapsedRealtime = elapsedRealtime; }
  public synchronized boolean register(String generation) {
    if (!identifier(generation) || sessions.containsKey(generation) || sessions.size() >= MAX_SESSIONS) return false;
    sessions.put(generation, new Session()); return true;
  }
  public synchronized boolean sessionActive(String generation) { return sessions.containsKey(generation); }
  public synchronized void revoke(String generation) {
    sessions.remove(generation);
    requests.entrySet().removeIf(entry -> entry.getValue().generation.equals(generation));
  }
  /** Snapshot is a complete JSON string constructed by the host from native registration and the original serialized request. */
  public synchronized String admit(String generation, long sequence, String snapshot) {
    prune();
    Session session = sessions.get(generation);
    if (session == null || sequence < 1 || sequence > MAX_SEQUENCE || sequence != session.sequence + 1) return null;
    // Consume each observed transport sequence, including overloads, so a rejected request cannot be replayed later.
    session.sequence = sequence;
    if (snapshot == null || snapshot.length() > MAX_BYTES || snapshot.getBytes(StandardCharsets.UTF_8).length > MAX_BYTES || requests.size() >= MAX_PENDING) return null;
    long ownPending = requests.values().stream().filter(request -> request.generation.equals(generation)).count();
    if (ownPending >= MAX_PER_SESSION) return null;
    long now = elapsedRealtime.getAsLong();
    if (now < 0 || now > Long.MAX_VALUE - LIFETIME_MS) return null;
    String token = UUID.randomUUID().toString();
    requests.put(token, new Request(generation, snapshot, now + LIFETIME_MS));
    return token;
  }
  public synchronized String take(String token) {
    prune();
    Request request = requests.get(token);
    if (request == null || request.claimed || !sessions.containsKey(request.generation)) return null;
    request.claimed = true; return request.snapshot;
  }
  public synchronized boolean isActive(String token) {
    prune();
    Request request = requests.get(token);
    return request != null && request.claimed && sessions.containsKey(request.generation);
  }
  /** Consumes exactly this request. Delivery must still recheck its generation on the UI thread. */
  public synchronized boolean finish(String token) {
    prune();
    Request request = requests.remove(token);
    return request != null && request.claimed && sessions.containsKey(request.generation);
  }
  private void prune() {
    long now = elapsedRealtime.getAsLong();
    requests.entrySet().removeIf(entry -> now < 0 || now >= entry.getValue().expires || !sessions.containsKey(entry.getValue().generation));
  }
  private static boolean identifier(String value) { return value != null && value.matches("[A-Za-z0-9_-]{1,80}"); }
}
